### PR TITLE
All controllers run by controller-manager use diffrent UserAgents

### LINKED ISSUE #17723: Make sure that every controller run by the controller manager uses separate client.
It's needed for finer-grained monitoring.

In the same time we need to make sure that all clients reuse the same transport to avoid creating unnecessary connections.

cc @lavalamp @brendandburns @wojtek-t @fgrzadkowski

### PR DESCRIPTION
**Pull request by gmarek**
Fixes #17723

cc @wojtek-t @lavalamp @krousey

This is a simple PR which creates a new client for every controller. This is inefficient, as we're creating different connections only to be able to pass a different UserAgent in the header. Sadly current implementation of transport does not make reusing it with different UserAgent simple. At the bottom there's a decorator which stores a UserAgent that is passed to it during transports creation and allowing changing it would require changes (read: breaking decorator pattern) in freshly rewritten transport library.

Ideas I have to do this either by dismantling the decorator, or by pulling UserAgentDecorator out of default constructor and require that users call it explicitly. Both are bad, so maybe @krousey have some better ideas.

### CODE REVIEW

**Comment by ncdc** [comment]
cc @kubernetes/rh-cluster-infra @liggitt @deads2k

**Comment by liggitt** [comment]
I actually want different clients per controller, so we can eventually scope their permissions

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 1]
```diff
@@ -335,34 +348,40 @@ func (s *CMServer) Run(_ []string) error {
 		glog.Infof("Starting %s apis", groupVersion)
 		if containsResource(resources, "horizontalpodautoscalers") {
 			glog.Infof("Starting horizontal pod controller.")
-			metricsClient := metrics.NewHeapsterMetricsClient(kubeClient, metrics.DefaultHeapsterNamespace, metrics.DefaultHeapsterScheme, metrics.DefaultHeapsterService, metrics.DefaultHeapsterPort)
-			podautoscaler.NewHorizontalController(kubeClient, metricsClient).
+			metricsClient := metrics.NewHeapsterMetricsClient(
+				defaultKubeClient,
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/controllermanager.go]: Give this a metrics client user agent?

Comment (gmarek) [inline cmd/kube-controller-manager/app/controllermanager.go]: Done.

**Comment by liggitt** [comment]
Do the QPS and burst defaults need to be revisited since clients are segmented now?

**Comment by gmarek** [comment]
It's a valid concern - I want to run experiments to see if we should, but today I was playing with huge clusters, so didn't have time to do so. I wouldn't worry about it too much before it turn out to be a problem (e.g. slow down our scalability tests).

**Comment by krousey** [comment]
@gmarek I'm technically out today, but wanted to address your concerns about different connections. TCP connections are pooled in each `http.Transport` object and reused there. If your client config doesn't specify any TLS config, you will keep reusing the default transport's pool with wrappers around it. If you do have TLS in your client config, then you will reuse and underlying cached transport with customer wrappers around it. https://github.com/kubernetes/kubernetes/blob/master/pkg/client/transport/transport.go#L43

Are you saying you're still somehow creating additional connections despite this layer of caching?

**Comment by gmarek** [comment]
@krousey - OK, I dug a little more around the code and I see what you mean: UserAgent is not used as a part of the key in transport cache, so it'll reuse the transport layer in all the clients I create. It's great - for me at least;)

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 2]
```diff
@@ -242,7 +252,7 @@ func (s *CMServer) Run(_ []string) error {
 	kubeconfig.QPS = s.KubeAPIQPS
 	kubeconfig.Burst = s.KubeAPIBurst
 
-	kubeClient, err := client.New(kubeconfig)
+	defaultKubeClient, err := client.New(kubeconfig)
```
Comment (krousey) [inline cmd/kube-controller-manager/app/controllermanager.go]: Minor nit: you didn't need to rename this.

**Comment by gmarek** [comment]
Addressed nit. Reapplying LTGM.

**Comment by gmarek** [comment]
@k8s-bot test this

**Comment by gmarek** [comment]
Trivial rebase

**Comment by gmarek** [comment]
@k8s-bot test this

**Pull request merged**